sqlalchemy-stubs/sql/elements.pyi

```diff
@@ -47,300 +58,378 @@ class ColumnElement(operators.ColumnOperators, ClauseElement):
     @classmethod
     def _instance(cls) -> True_: ...
 
+_CL = TypeVar('_CL', bound=ClauseList)
+
 class ClauseList(ClauseElement):
     __visit_name__: str = ...
     operator: Any = ...
-    group: Any = ...
-    group_contents: Any = ...
-    clauses: Any = ...
-    def __init__(self, *clauses, **kwargs) -> None: ...
-    def __iter__(self): ...
-    def __len__(self): ...
-    def append(self, clause): ...
-    def get_children(self, **kwargs): ...
-    def self_group(self, against: Optional[Any] = ...): ...
-    def compare(self, other, **kw): ...
-
-class BooleanClauseList(ClauseList, ColumnElement):
+    group: bool = ...
+    group_contents: bool = ...
+    clauses: List[ClauseElement] = ...
+    def __init__(self, *clauses: ClauseElement, operator: Callable[..., Any] = ..., group: bool = ...,
+                 group_contents: bool = ..., **kwargs: Any) -> None: ...
+    def __iter__(self) -> Iterator[ClauseElement]: ...
+    def __len__(self) -> int: ...
+    def append(self, clause: ClauseElement): ...
+    def get_children(self, **kwargs) -> List[ClauseElement]: ...
+    def self_group(self: _CL, against: Optional[Any] = ...) -> Union[_CL, Grouping[_T]]: ...
+    def compare(self, other: Any, **kw: Any) -> bool: ...
+
+_BCL = TypeVar('_BCL', bound=BooleanClauseList)
+
+class BooleanClauseList(ClauseList, ColumnElement[Boolean]):
     __visit_name__: str = ...
-    def __init__(self, *arg, **kw) -> None: ...
     @classmethod
-    def and_(cls, *clauses) -> BooleanClauseList: ...
+    def and_(cls, *clauses: ClauseElement) -> BooleanClauseList: ...
     @classmethod
-    def or_(cls, *clauses) -> BooleanClauseList: ...
-    def self_group(self, against: Optional[Any] = ...): ...
+    def or_(cls, *clauses: ClauseElement) -> BooleanClauseList: ...
+    def self_group(self: _BCL, against: Optional[Any] = ...) -> Union[_BCL, Grouping[_T]]: ...
 
 and_ = BooleanClauseList.and_
 or_ = BooleanClauseList.or_
 
-class Tuple(ClauseList, ColumnElement):
-    type: Any = ...
-    def __init__(self, *clauses, **kw) -> None: ...
+class Tuple(ClauseList, ColumnElement[_T]):
+    type: _T = ...
+    @overload
+    def __init__(self, *clauses: ColumnElement[Any], type_: Type[_T], **kw: Any) -> None: ...
+    @overload
+    def __init__(self, *clauses: ColumnElement[Any], type_: _T, **kw: Any) -> None: ...
+    @overload
+    def __init__(self, clause: ColumnElement[_T], *clauses: ColumnElement[Any], **kw: Any) -> None: ...
 
-class Case(ColumnElement):
+class Case(ColumnElement[TypeEngine[_T]], Generic[_T]):
     __visit_name__: str = ...
     value: Any = ...
-    type: Any = ...
+    type: TypeEngine[_T] = ...
     whens: Any = ...
     else_: Any = ...
-    def __init__(self, whens, value: Optional[Any] = ..., else_: Optional[Any] = ...) -> None: ...
-    def get_children(self, **kwargs): ...
+    def __init__(self, whens: List[_TupleType[ClauseElement, _T]], value: Optional[_T] = ...,
```

Comment:
Also note that many APIs (I am not sure about SQLAlchemy) that accept `Iterable[Tuple[A, B]]` also accept `Dict[A, B]`.